### PR TITLE
Concurrency: do not remove shared or dirty pointers from the value set

### PR DESCRIPTION
**Pull request by tautschnig**
Just like we always add (and never replace) shared pointers in
goto_symex_statet::assignment, symex_dead should not remove shared or
dirty pointers.

This fixes a first problem, with further problems remaining as
indicated by the KNOWNBUG regression test.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by martin-cs — APPROVED** [review]
Approve but if you have time to look at the `KNOWNBUG` before merging, that would be great.

**Comment by tautschnig** [comment]
> Approve but if you have time to look at the `KNOWNBUG` before merging, that would be great.

Oh, my apologies for failing to make that clear: that's fixed in #6122.

**Review by kroening — APPROVED** [review]

**Pull request merged**
